### PR TITLE
add functions to convert shape_msgs/SolidPrimitive <-> euslisp object

### PR DESCRIPTION
**Pull request by 708yamaguchi**
I added functions to convert `shape_msgs/SolidPrimitive` <-> `euslisp object`

`shape_msgs/SolidPrimitive` is used at collision-object-publisher.l, for example.
https://github.com/jsk-ros-pkg/jsk_pr2eus/blob/master/pr2eus_moveit/euslisp/collision-object-publisher.l#L80-L85

If this pull request is merged, I will replace the above code with cube->shape-msg() functions.

### CODE REVIEW

File: roseus/euslisp/roseus-utils.l [hunk 1]
```diff
@@ -620,6 +620,75 @@
...
+     ((eq type shape_msgs::SolidPrimitive::*BOX*)
+      (shape-msg->shape/cube msg))
+     ((eq type shape_msgs::SolidPrimitive::*SPHERE*)
+      (shape-msg->shape/sphere msg))
+     ((eq type shape_msgs::SolidPrimitive::*CYLINDER*)
+      (shape-msg->shape/cylinder msg))
+     (t
+      (error "unknown type ~A" type)))))
+
+(defun shape-msg->shape/cube (msg)
```
Comment (knorth55) [inline roseus/euslisp/roseus-utils.l:670]: how about `shape-msg->cube` ?

File: roseus/euslisp/roseus-utils.l [hunk 2]
```diff
@@ -620,6 +620,75 @@
...
+  "Convert shape_msgs::SolidPrimitive to euslisp cube object"
+  (let* ((scale (send msg :dimensions))
+         (cb (make-cube (* (elt scale 0) 1000)
+                        (* (elt scale 1) 1000)
+                        (* (elt scale 2) 1000))))
+    cb))
+
+(defun shape-msg->shape/sphere (msg)
```
Comment (knorth55) [inline roseus/euslisp/roseus-utils.l:678]: how about `shape-msg->sphere` ?

File: roseus/euslisp/roseus-utils.l [hunk 3]
```diff
@@ -620,6 +620,75 @@
...
+  "Convert shape_msgs::SolidPrimitive to euslisp sphere object"
+  (let* ((scale (send msg :dimensions))
+         (cp (make-sphere (* (elt scale 0) 1000))))
+    cp))
+
+(defun shape-msg->shape/cylinder (msg)
```
Comment (knorth55) [inline roseus/euslisp/roseus-utils.l:684]: how about `shape-msg->cylinder` ?

**Comment by 708yamaguchi** [comment]
Thank you for your review.

I named these functions following existing functions:
https://github.com/jsk-ros-pkg/jsk_roseus/blob/master/roseus/euslisp/roseus-utils.l#L537
https://github.com/jsk-ros-pkg/jsk_roseus/blob/master/roseus/euslisp/roseus-utils.l#L550
https://github.com/jsk-ros-pkg/jsk_roseus/blob/master/roseus/euslisp/roseus-utils.l#L567

I pushed an additional commit to reflect your advice.

**Comment by k-okada** [comment]
please check travis errors
```

[roseus:make] ;; loading roseus("1.7.4-54-g23fdb92") on euslisp((9.27 ip-172-30-1-167 Wed Mar 4 01:33:18 UTC 2020  1.2.1))
5300[roseus:make] Traceback (most recent call last):
5301[roseus:make]   File "/opt/ros/melodic/share/rostest/nodes/hztest", line 243, in <module>
5302[roseus:make]     time.sleep(0.75)
5303[roseus:make] KeyboardInterrupt
5304[roseus:make] [ERROR] [1586526159.057168430]: Could not find roseus messages for shape_msgs under (/home/travis/ros/ws_jsk_roseus/devel/.private/roseus/share /home/travis/ros/ws_jsk_roseus/devel/share /opt/ros/melodic/share)
5305[roseus:make] try rosrun roseus generate-all-msg-srv.sh shape_msgs
5306[roseus:make] ;; can not find nil/msg/ directory for [shape_msgs] packages
5307[roseus:make] Built target _run_tests_roseus_rostest_test_test-simple-client-wait.test
5308[roseus:make] Scanning dependencies of target _run_tests_roseus_rostest_test_test-namespace.test
5309[roseus:make] -- run_tests.py: execute commands

```

**Comment by 708yamaguchi** [comment]
Thank you for your advice.

### Cause of error

I have checked the travis error carefully, I found following:

- kinetic and melodic tests
  In success tests of kinetic and melodic, shape_msgs is installed via dependency of jsk_pcl_ros.
  In failure tests of kinetic and melodic, shape_msgs is not installed because jsk_pcl_ros cannot be installed.
  ```
  E: Unable to locate package ros-melodic-jsk-pcl-ros
  ```
  https://api.travis-ci.org/v3/job/673356522/log.txt

  This difference might be caused by $ROS_REPOSITORY_PATH. Only when we use http://packages.ros.org/ros-testing/ubuntu in melodic, ros-melodic-jsk-pcl-ros is not found.
  I checked this in my PC. I seach for ros-melodic-jsk-pcl-ros on `deb http://packages.ros.org/ros-testing/ubuntu bionic main`, but it is not found.
  ```
  $ sudo apt update
  ...
  $ apt search ros-melodic-jsk-pcl-ros
  Sorting... Done
  Full Text Search... Done
  ros-melodic-jsk-pcl-ros-utils/bionic 1.2.10-0bionic.20200406.144533 amd64 [upgradable from: 1.2.10-0bionic.20200320.164245]
  ROS utility nodelets for pointcloud perception.

  ros-melodic-jsk-pcl-ros-utils-dbgsym/bionic 1.2.10-0bionic.20200406.144533 amd64
  ```

- indigo tests
  In all tests, shape_msgs is installed via dependency of jsk_pcl_ros.

  However, All of the failed tests say:
  ```
  Setting up ros-indigo-shape-msgs (1.11.10-0trusty-20190604-085849-0800) ...
  ...
  [0m[31m;; can not find :no-msg-package/msg/ directory for [shape_msgs] packages
  ```
  and\
  All of the failed tests except 'TEST_PKGS="task_compiler"' say:
  ```
  Summary: 45 tests, 0 errors, 45 failures, 0 skipped
  ```

  I think the reason is manifest.l of shape_msgs is not created in indigo environment, but I cannot understand why manifest.l is not created while ros-indigo-shape-msgs is installed...
  https://github.com/jsk-ros-pkg/jsk_roseus/blob/master/roseus/euslisp/roseus.l#L406

### What I should do
For kinetic and melodic tests, I think we should wait for jsk_pcl_ros to be released on http://packages.ros.org/ros-testing/ubuntu
For indigo tests, I have no idea. Do I need to add shape_msgs to run_depend in roseus? Or is there any way to avoid the above error?
Could you please teach me how to fix these problem?

**Comment by knorth55** [comment]
this issue is related to https://github.com/jsk-ros-pkg/jsk_visualization/pull/765 .
jsk_pcl_ros is not released because jsk_visualization build failed .

**Comment by k-okada** [comment]
Please add depends to `shape_msgs` from `roseus` as I noted on https://github.com/jsk-ros-pkg/jsk_roseus/pull/641#issuecomment-611292100. 
we do not want `roseus` depends on `jsk_pcl_ros`, whcih is large dependencies. please also note that `shape_msgs` provided from `jsk_pcl_ros` of `roseus_tutorials`. But this does not work when building deb file, becuase deb file created for each package, not each repository. So when system create `roseus` package, it doens not know about dependencies of `roseus_tutorials`

**Comment by k-okada** [comment]
please refer https://github.com/jsk-ros-pkg/jsk_visualization/pull/699#issuecomment-410621239 https://github.com/jsk-ros-pkg/jsk_visualization/pull/644#issuecomment-262473691 for jsk_visualizaiotn case. this is more complex than jsk_roseus, but it seems we do not want jsk_visualization depends on jsk_recognition.

**Comment by 708yamaguchi** [comment]
Thank you for your teaching.

>  But this does not work when building deb file, becuase deb file created for each package, not each repository. So when system create roseus package, it doens not know about dependencies of roseus_tutorials

I have misunderstood the package system.
Now I understand why we need to `shape_msgs` to `roseus` dependency.

Also, from these comments (https://github.com/jsk-ros-pkg/jsk_roseus/pull/641#issuecomment-611292100, https://github.com/jsk-ros-pkg/jsk_visualization/pull/699#issuecomment-410621239 and https://github.com/jsk-ros-pkg/jsk_visualization/pull/644#issuecomment-262473691), I understand that we should not add large packages dependencies.
I'll be careful from now on.

**Comment by knorth55** [comment]
what  i waant to tell you is that jsk_pcl_ros is not released in ros-testing is because of jsk_visualization.
maybe it is good time to learn about ros buildfarm.
https://github.com/jsk-ros-pkg/jsk_visualization/issues/764

**Comment by 708yamaguchi** [comment]
@knorth55 
Thank you very much for your information.

Now, I am trying to understand the system of ROS Buildfarm.
e.g. xx-release repo, bloom command, jenkins, ros-shadow-fixed, ... etc

In my understanding, the current situation is as follows:
1. `jsk_rviz_plugins` build on build.ros.org fails because some dependent packages have been changed.
  http://build.ros.org/job/Mbin_uB64__jsk_rviz_plugins__ubuntu_bionic_amd64__binary/
1. `jsk_interactive_marker` build on build.ros.org (which depends on `jsk_rviz_plugins`) is pending.
  http://build.ros.org/job/Mbin_uB64__jsk_interactive_marker__ubuntu_bionic_amd64__binary/
1. `jsk_pcl_ros` build on build.ros.org (which depends on `jsk_interactive_marker`) is pending.
  http://build.ros.org/job/Mbin_uB64__jsk_pcl_ros__ubuntu_bionic_amd64__binary/
1. `jsk_pcl_ros` is not released on ros-testing (the same as ros-shadow-fixed?)

**Comment by pazeshun** [comment]
`ros-shadow-fixed` is old repository of `ros-testing`:
https://discourse.ros.org/t/security-issue-on-ros-build-farm/9342/8

**Comment by 708yamaguchi** [comment]
@pazeshun 
Thank you for correction!

**Pull request merged**